src/Misc/PerformanceViewer/performanceViewerCollector.ts

```diff
@@ -0,0 +1,208 @@
+
+            this._datasetMeta.set(strategy.id, {
+                color: this._getHexFromId(strategy.id),
+            });
+
+            this._strategies.set(strategy.id, strategy);
+        }
+
+        this.metadataObservable.notifyObservers(this._datasetMeta);
+    }
+
+    /**
+     * Gets a 6 character hexcode from a passed in string.
+     * @param id the string to get a hex code for.
+     * @returns a hexcode hashed from the id.
+     */
+    private _getHexFromId(id: string) {
+        // this first bit is just a known way of hashing a string.
+        let hash = 0;
+        for (let i = 0; i < id.length; i++) {
+            // (hash << 5) - hash is the same as hash * 31
+            hash = id.charCodeAt(i) + ((hash << 5) - hash);
+        }
+
+        // then we build the string octet by octet.
+        let hex = "#";
+        for (let i = 0; i < numberOfBitsInHexcode; i += 8) {
+            const octet = (hash >> i) & 0xFF;
+            hex += (hexPadding + octet.toString(16)).substr(-2);
+        }
+
+        return hex;
+    }
+
+    /**
+     * Collects data for every dataset by using the appropriate strategy. This is called every frame.
+     * This method will then notify all observers with the latest slice.
+     */
+    private _collectDataAtFrame = () => {
+        const timestamp = PrecisionDate.Now - this._startingTimestamp;
+        const numPoints = this.datasets.ids.length;
+
+        // add the starting index for the slice
+        const numberOfIndices = this.datasets.startingIndices.itemLength;
+        let startingIndex = 0;
+
+        if (numberOfIndices > 0) {
+            const previousStartingIndex = this.datasets.startingIndices.at(numberOfIndices - 1);
+            startingIndex = previousStartingIndex + this.datasets.data.at(previousStartingIndex + numberOfPointsOffset) + sliceDataOffset;
+        }
+
+        this.datasets.startingIndices.push(startingIndex);
+
+        // add the first 2 items in our slice.
+        this.datasets.data.push(timestamp);
+        this.datasets.data.push(numPoints);
+
+        // add the values inside the slice.
+        this.datasets.ids.forEach((id: string) => {
+            const strategy = this._strategies.get(id);
+
+            if (!strategy) {
+                return;
+            }
+
+            this.datasets.data.push(strategy.getData());
+        });
+
+        if (this.datasetObservable.hasObservers()) {
+            const slice: number[] = [timestamp, numPoints];
+
+            for (let i = 0; i < numPoints; i++) {
+                slice.push(this.datasets.data.at(startingIndex + sliceDataOffset + i));
+            }
+
+            this.datasetObservable.notifyObservers({slice});
+        }
+    }
+
+    /**
+     * Collects data for every dataset by using the appropriate strategy when the user wants.
+     * This method will then notify all observers with the latest slice of data.
+     */
+    public collectDataNow() {

```

Comment:
I might be replacing this method with a single public collectData (which will just be _collectDataAtFrame renamed) if it makes sense to also update the datasets when the user calls this function.

Reply:
What I have right now seems to be fine.